op-node/rollup/derive/channel_out.go

```diff
@@ -99,40 +125,119 @@ func (co *ChannelOut) AddBlock(block *types.Block) (uint64, error) {
 		return 0, errors.New("already closed")
 	}
 
-	batch, _, err := BlockToBatch(block)
+	batch, _, err := BlockToSingularBatch(block)
 	if err != nil {
 		return 0, err
 	}
-	return co.AddBatch(batch)
+	return co.AddSingularBatch(batch)
 }
 
-// AddBatch adds a batch to the channel. It returns the RLP encoded byte size
+// AddSingularBatch adds a batch to the channel. It returns the RLP encoded byte size
 // and an error if there is a problem adding the batch. The only sentinel error
 // that it returns is ErrTooManyRLPBytes. If this error is returned, the channel
 // should be closed and a new one should be made.
 //
-// AddBatch should be used together with BlockToBatch if you need to access the
+// AddSingularBatch should be used together with BlockToSingularBatch if you need to access the
 // BatchData before adding a block to the channel. It isn't possible to access
 // the batch data with AddBlock.
-func (co *ChannelOut) AddBatch(batch *BatchData) (uint64, error) {
+func (co *ChannelOut) AddSingularBatch(batch *SingularBatch) (uint64, error) {
 	if co.closed {
 		return 0, errors.New("already closed")
 	}
 
+	switch co.batchType {
+	case SingularBatchType:
+		return co.writeSingularBatch(batch)
+	case SpanBatchType:
+		return co.writeSpanBatch(batch)
+	default:
+		return 0, fmt.Errorf("unrecognized batch type: %d", co.batchType)
+	}
+}
+
+func (co *ChannelOut) writeSingularBatch(batch *SingularBatch) (uint64, error) {
+	var buf bytes.Buffer
 	// We encode to a temporary buffer to determine the encoded length to
 	// ensure that the total size of all RLP elements is less than or equal to MAX_RLP_BYTES_PER_CHANNEL
-	var buf bytes.Buffer
-	if err := rlp.Encode(&buf, batch); err != nil {
+	if err := rlp.Encode(&buf, NewSingularBatchData(*batch)); err != nil {
 		return 0, err
 	}
 	if co.rlpLength+buf.Len() > MaxRLPBytesPerChannel {
-		return 0, fmt.Errorf("could not add %d bytes to channel of %d bytes, max is %d. err: %w",
+		co.fullErr = fmt.Errorf("could not add %d bytes to channel of %d bytes, max is %d. err: %w",
 			buf.Len(), co.rlpLength, MaxRLPBytesPerChannel, ErrTooManyRLPBytes)
+		return 0, co.fullErr
 	}
 	co.rlpLength += buf.Len()
 
 	// avoid using io.Copy here, because we need all or nothing
 	written, err := co.compress.Write(buf.Bytes())
+	if co.compress.FullErr() != nil {
+		co.fullErr = co.compress.FullErr()
+	}
+	return uint64(written), err
+}
+
+// writeSpanBatch appends a SingularBatch to the channel's SpanBatch.
+// A channel can have only one SpanBatch. And compressed results should not be accessible until the channel is closed, since the prefix can be changed.
```

Comment:
```suggestion
// A channel can have only one SpanBatch. And compressed results should not be accessible until the channel is closed, since the prefix or payload can be changed.
```